core/src/main/java/com/linecorp/armeria/server/ServiceConfigBuilder.java

```diff
@@ -343,33 +341,48 @@ ServiceConfig build(ServiceNaming defaultServiceNaming,
                                                                      unhandledExceptionsReporter);
         }
 
-        final boolean webSocket = service.as(DefaultWebSocketService.class) != null;
+        final HttpServiceOptions options = service.options();
```

Comment:
Should we apply the properties of `HttpServiceOptions` in the constructor of this class?
We may change the primitive `long` into `Long` to know if a value is not specified.
The values of `HttpServiceOptions` could be overridden by the setters of this builder.

Afterwards, the following code can be simplified.
```java
final long requestTimeoutMillis;
if (this.requestTimeoutMillis != null) {
  requestTimeoutMillis = this.requestTimeoutMillis;
} else {
  requestTimeoutMillis = Flags.defaultRequestTimeoutMillis();
}
...
```

Reply:
The code was modified so that the properties of `HttpServiceOptions` can be applied in the constructor of `ServiceConfigBuilder`.

The order of priority for setting the fields is as follows:
1. setters
2. constructor
3. default values(passed by `ServiceConfigBuilder#build`) 

If the constructor has set a field, default values passed from the `ServiceConfigBuilder#build` will not be applied due to the lower priority. As a result, `WebSocketServiceConfigTest#webSocketServiceDefaultConfigValues` would fail. Should we allow default `HttpServiceOptions` to override the values passed by the `ServiceConfigBuilder#build`? (A comment has been added to the `WebSocketServiceConfigTest` for further explanation).

Reply:
> Should we allow default HttpServiceOptions to have higer priority than the values passed by the ServiceConfigBuilder#build? 

No, I don't think so. `default` means the value will be used when no custom option isn't set.
1. Flags (lowest)
2. HttpServiceOptions 
3. Builders (highest)

Reply:
I understood the problem you meant. Please ignore the previous comment about applying `HttpServiceOptions` to the fields in the constructor.

How about passing nullable values to `ServiceConfigBuilder and not falling back to ` template.requestTimeoutMillis`?

```java
long maxRequestLength = this.maxRequestLength;
if (maxRequestLength == null) {
   maxRequestLength = defaultRequestTimeoutMillis;
}
if (maxRequestLength) {
   maxRequestLength = options.maxRequestLength();
}
```
To use this approach, the default value of HttpServiceOptions should not be nullable. We should set them with the value of `Flags`.
```java
public class HttpServiceOptionsBuilder {
    private long requestTimeoutMillis = Flags.defaultRequestTimeoutMillis();
    ...
```

Reply:
I had a chat with other maintainers about the priorities of `HttpServiceOptions`. The new direction differs from my previous comments. Let me resolve this conversation and start a new one.